spark-operator/src/main/java/org/apache/spark/k8s/operator/metrics/MetricsSystem.java

```diff
@@ -0,0 +1,139 @@
+ * under the License.
+ */
+
+package org.apache.spark.k8s.operator.metrics;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import com.codahale.metrics.MetricFilter;
+import com.codahale.metrics.MetricRegistry;
+import lombok.Data;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.spark.k8s.operator.metrics.source.OperatorJvmSource;
+import org.apache.spark.metrics.sink.Sink;
+import org.apache.spark.metrics.source.Source;
+
+@Slf4j
+public class MetricsSystem {
+  private final AtomicBoolean running = new AtomicBoolean(false);
+  @Getter private final Set<Sink> sinks;
+  @Getter private final Set<Source> sources;
+  @Getter private final MetricRegistry registry;
+  @Getter private final Properties properties;
+  // PrometheusPullModelHandler is registered by default, metrics exposed via http port
+  @Getter private final PrometheusPullModelHandler prometheusPullModelHandler;
+  private final Map<String, SinkProps> sinkPropertiesMap;
+
+  public MetricsSystem() {
+    this(new Properties());
+  }
+
+  public MetricsSystem(Properties properties) {
+    this.sources = new HashSet<>();
+    this.sinks = new HashSet<>();
+    this.registry = new MetricRegistry();
+    this.properties = properties;
+    this.sinkPropertiesMap = MetricsSystemFactory.parseSinkProperties(this.properties);
+    // Add default sinks
+    this.prometheusPullModelHandler = new PrometheusPullModelHandler(new Properties(), registry);
+    this.sinks.add(prometheusPullModelHandler);
+  }
+
+  public void start() {
+    if (running.get()) {
+      throw new IllegalStateException(
+          "Attempting to start a MetricsSystem that is already running");
+    }
+    running.set(true);
+    registerDefaultSources();
+    registerSinks();
+    sinks.forEach(Sink::start);
+  }
+
+  public void stop() {
+    if (running.get()) {
+      sinks.forEach(Sink::stop);
+      registry.removeMatching(MetricFilter.ALL);
+    } else {
+      log.error("Stopping a MetricsSystem that is not running");
+    }
+    running.set(false);
+  }
+
+  public void report() {
+    sinks.forEach(Sink::report);
+  }
+
+  protected void registerDefaultSources() {
+    registerSource(new OperatorJvmSource());
+  }
+
+  protected void registerSinks() {
+    log.info("sinkPropertiesMap: {}", sinkPropertiesMap);
+    sinkPropertiesMap
+        .values()
+        .forEach(
+            sinkProp -> {
+              try {
+                Class<Sink> sinkClass = (Class<Sink>) Class.forName(sinkProp.getClassName());
+                Sink sinkInstance;
+                sinkInstance =
+                    sinkClass
+                        .getConstructor(Properties.class, MetricRegistry.class)
+                        .newInstance(sinkProp.getProperties(), registry);
+                sinks.add(sinkInstance);
+              } catch (Exception e) {
```

Comment:
Is this inevitable? Otherwise, please always try to use the narrowest exception time for readability.